self-hosted spack, or container spackOS

trws opened the pull request:
This is a big, messy, WIP PR right now but collects all the changes I made to get spack to bootstrap a toolchain from glibc to a fully functional gcc, python and spack environment that can be relocated independently of the base system.  An example container for x86_64 containing nothing but spack-built software in a spack root at /spack (and a couple of text files and symlinks) that is also a working bootstrapped spack installation and userland can be found at "trws/spackos-experimental:spackos" on dockerhub. For aarch64 use trws/spackos-experimental:spackos-aarch64

Things that work surprisingly well:
* Relocation just worked.  *ALL THE WAY DOWN*.  I'm pleasantly gobsmacked, and built the entire stack in one path, relocated it to another in the container and stuff just works.
* Because of 1, we can now build spack binary packages with dynamically linked executables that will work on any linux distro.  Download them on a musl system, and they just go.
* After slapping in a couple of symlinks to provide `/bin/sh` and `/usr/bin/env` from the spack env at `/spack/base` and adding super-basic `passwd` and `shadow` files, the base system in the container pretty much just works.  Bash, coreutils, basically everything you would expect, and despite spack bootstrap claiming it doesn't have clingo on first entry, a "spack bootstrap now" cleans that up without actually installing anything. 🤷

Things that definitely need work:
* Actually doing the bootstrap is a PITA right now, and could be substantially more automated.  Current process:
  * On an ubuntu22.04 x86_64 system (yes, this is a requirement for now) activate each environment in `bootstraps/stage{1-4}` in sequence and build them with the following caveats.  To make a "spackos" container, delete the `.git/*` line from the `.dockerignore` file, then use the dockerfile in the bootstraps folder with this command from the spack root: `docker buildx build -t spackos-experimental --load -f bootstraps/Dockerfile --target=spackos .`
  * In stage4 the solver _really_ likes to build a second glibc, and because the gcc includes one by default, you end up mixing two glibc instances, which ends very, very badly.  Before building stage4 the hash for the stage2/3 glibc has to be _forced_ in the env file for stage4.
  * Also in stage4, the python build does something magical that pulls in headers from /usr/local/* or /usr/include/x86_64-linux-gnu for some reason.  Currently work around this with bubblewrap like so: `bwrap --dev-bind / / --tmpfs /usr/local --tmpfs /usr/include/x86_64-linux-gnu spack install -v python`
  * Many stage1 items are not strictly "necessary" and can be externals, but if externals are used _anywhere_ in here then the ubuntu and spack OS requirements get messed up and it becomes nearly impossible to build anything useful in stage3 or later.  Best to just build a bunch of extra crud for now.
* Spack currently doesn't really know what the "spack" OS is, and hasn't grown a mechanism to handle OS switches like that yet, so it has to be overridden all over the place.
* Base system libraries need special handling to keep the wrapper from mangling them.  If glibc is added to the include path with `-isystem`, C++ compilation will fail because the glibc headers will be moved before the libstdc++ ones.  The current workaround for that is really nasty.
* The whole thing currently only works by lying to spack and saying that glibc and libxcrypt have no dependencies... This is not good, but it works until we can work around the issues in concretizing the multiple copies of all their build deps and avoid having them stick around in the graph.
* Spack currently doesn't provide a mechanism for a "durable" environment (a way to keep an attempted rebuild from killing the currently working env if it fails) or pinning a version of a base package like glibc other than very explicitly pinning the hash.  This makes the experience of using it to manage everything a little rough.  I would actually recommend playing with it to see how it works/feels, we'll need something new there for sure.

trws committed:
in-progress bootstrap setup

trws committed:
add libstdcxx

trws committed:
first successful gcc

trws committed:
allow setting a spackos through env for now

trws committed:
better gcc flag handling for the stage1 compiler

trws committed:
cc: work around -v split between ld and ccld

trws committed:
almost complete stage 3 gcc, should be fully independent now

trws committed:
gcc: inject -B into configure to build runtime libs

trws committed:
hide build deps from glibc, nasty but necessary

trws committed:
ensure stage1 does not try to build with threads

trws committed:
fix python library searching

trws committed:
tweak env, fix linux-pam build

trws committed:
ensure headers are copied into place in glibc

trws committed:
patch system paths out of gcc when built for spackos

trws committed:
first version of spack stage environment files

trws committed:
switch sysroot variant to stage0 since we do not need the value

trws committed:
stage0 refactor complete, successful build

trws committed:
linux-headers needs to be NoDep

trws committed:
refactor up through stage2

trws committed:
fix up glibc/libxcrypt include handling to use idirafter

trws committed:
version 2 full bootstrap

trws committed:
fix broken stage4 path for docker, add dockerfile

trws committed:
mark more things as build-tools

trws committed:
tweak glibc deps

trws committed:
avoid glibc cycle through gcc-runtime

trws committed:
avoid python using system paths

trws committed:
munge only installs a pkgconfig if /usr/lib/pkgconfig or similar exists, fix

trws committed:
add some bootstrap automation, use docker to make it reasonably reproducible

trws committed:
refactor bootstrap to work on multiple architectures

trws committed:
add padding and fix ARCH replacement

trws committed:
make it easier to drop into the docker env

trws committed:
break cycle for gnuconfig dep

trws committed:
gcc needs crypt for libsanitizer

trws committed:
fix bootstrap

trws committed:
do not use +bootstrap on gcc, too fragile

haampie commented:
Few comments:

* We could use this approach https://github.com/haampie/spack-intermediate-gcc-example to express the dependencies between the different bootstrap stages, and simplify configuration, so that bootstrapping is a single `make` invocation.
* You're using Docker just to get a toolchain to work with, right? But in principle it can also be the user's own toolchain?

haampie commented on bootstraps/stage1/spack.in.yaml:
```diff
@@ -0,0 +1,28 @@
+  concretizer:
+    unify: true
+
+  compilers:
```
Preferably `spack -e stage1 compiler find`

trws replied:
Now that we’re building up a driving script I think I agree. This was the solution to making as much of it as possible defined in the environment. Either way it needs to be in the env, because we need to control which gcc@13 is selected, and it’s the only way I’ve found.

haampie commented on bootstraps/stage1/spack.in.yaml:
```diff
@@ -0,0 +1,28 @@
+spack:
+  specs:
+  - binutils +stage0
+  - gcc +stage0 +binutils ~bootstrap languages=c,c++
```
Should `gcc@13.2.0` be pinned here?

trws replied:
Because of the other envs yes it should, good point.

haampie commented on bootstraps/stage2/spack.in.yaml:
```diff
@@ -0,0 +1,52 @@
+spack:
+  # add package specs to the `specs` list
+  specs:
+  - glibc %gcc@=13.2.0 arch=linux-spack-x86_64 ^linux-headers%gcc@12
```
Can `%gcc@12` be dropped from `linux-headers` or is it installing stuff conditionally after inspecting the compiler?

trws replied:
Linux-headers needs a fully-functional compiler, or at least a compiler with a libc, to build. It doesn’t seem to matter if it’s the host or target compiler though, so this is my solution to getting it to use the host compiler for just that one package. The version isn’t important, again it’s the only way I know to pin a specific compiler.

I’m now considering whether I could script it so that we generate a larger number of env files, and just generate an env whenever we want to change compilers to get rid of these version pins. Should work, but it will be more pieces.

trws replied:
I switched this to using a requirement for a compiler version lower than the one we're compiling.  I don't like that solution, but nothing else I've tried has worked.

haampie commented on bootstraps/stage2/spack.in.yaml:
```diff
@@ -0,0 +1,52 @@
+    unify: true
+  packages:
+    all:
+      target: [x86_64]
```
Those `packages:all:target` should probably be changed to `packages:require:[target=x86_64,aarch64]` (`target` accepts a list of options)

trws replied:
Interesting, I’ll try that thanks.

trws replied:
Worked great!

haampie commented on bootstraps/stage2/spack.in.yaml:
```diff
@@ -0,0 +1,52 @@
+        fc: null
+      flags: {}
+      operating_system: spack
+      target: x86_64
```
And `target: x86_64` is I think not required, but if it is required, then we should rely on `spack compiler find` to fill out the details.

haampie commented on bootstraps/stage3/spack.in.yaml:
```diff
@@ -0,0 +1,48 @@
+      modules: []
+      environment: {}
+      extra_rpaths: []
+  - compiler:
```
Remove?

trws replied:
Ah, that was there so I could use depfile on stage3 to save build time. Yeah, it can come out.

trws replied:
This is now gone.

haampie commented on bootstraps/stage4/spack.in.yaml:
```diff
@@ -0,0 +1,74 @@
+  - binutils +gold +lto arch=linux-spack-x86_64
+  - gcc +binutils +piclibs ~bootstrap languages=c,c++,fortran,lto arch=linux-spack-x86_64
+  # base system
+  - m4 arch=linux-spack-x86_64
```
Drop all hard-coded `arch=linux-spack-=x86_64` in favor of a single more flexible requirement `target=x86_64,aarch64 os=spack`

trws replied:
Will see if that works, last I tried applying os=spack that way did not work. Might with the override now that I think about it, good thought.

trws replied:
The requirement works.  The preferences interface I was using before didn't, this helped clean up a lot of messy complexity.

haampie commented:
I did an initial pass over just the bootstrapping envs, will look more into how you deal with newlib.

I think the `newlib` fits Spack's model better as it doesn't require concurrent staging of `gcc` and `glibc`. However the other approach* is faster: 

\*: (1) install glibc and linux-kernel headers (2) configure GCC once and for all, make just xgcc w/ libgcc.a (3) use that to build glibc (4) build the rest of gcc (with -B flag). [More details](https://github.com/spack/spack/wiki/Building-gcc-with-glibc-together-in-10-easy-steps)

haampie commented on var/spack/repos/builtin/packages/glibc/package.py:
```diff
@@ -155,22 +199,25 @@ def patch(self):
+        depends_on("python@3.4:", type="build", when="@2.29:")
 
-    depends_on("linux-headers")
+    depends_on("linux-headers", type="build")
```
This was `link` type before s.t. you'd get automatic `-I` flags. Can you leave a comment that this is no longer necessary cause the headers are copied?

Note that it could be somewhat problematic with reuse concretization, since `linux-headers` would be pruned if pure build dep, meaning people can't put conflicts on it (for whatever reason) or figure out easily what version of the headers they have?

trws replied:
Hm, that’s a good point. I can definitely add the comment at least. Worth some testing.

haampie commented on bootstraps/stage1/spack.in.yaml:
```diff
@@ -0,0 +1,28 @@
+# configuration settings.
+spack:
+  specs:
+  - binutils +stage0
```
`stage1/spack.yaml` with `+stage0` is confusing ;p

trws replied:
I agree, _sigh_ iterative naming is not good naming.

haampie commented on bootstraps/stage2/spack.in.yaml:
```diff
@@ -0,0 +1,52 @@
+  - glibc %gcc@=13.2.0 arch=linux-spack-x86_64 ^linux-headers%gcc@12
+  - binutils %gcc@=13.2.0 arch=linux-spack-x86_64
+  - gcc@13.2.0 %gcc@=13.2.0 ~bootstrap ~graphite +binutils arch=linux-spack-x86_64
+  - gmp +stage1 arch=linux-spack-x86_64
```
`gmp` would also be built with stage0 gcc, right? Maybe be explicit?

trws replied:
Probably best yeah.

haampie commented on var/spack/repos/builtin/packages/gmp/package.py:
```diff
@@ -62,4 +72,22 @@ def configure_args(self):
+                '-isystem ' ,
+                glibc.prefix.include,
+                '--sysroot=/',
+                '-B ', glibc.prefix,
```
I may be mistaken but I remember white-space between `-B /path` is either not allowed or leads to issues with libtool and friends. Maybe `f"-B{prefix}"` to be on the safe side?

trws replied:
Sure. FWIW, it can cause issues with libtool, but we are forced to work around libtool anyway because it likes to _silently ignore_ -B flags (with or without space) in some situations.

haampie commented on var/spack/repos/builtin/packages/gmp/package.py:
```diff
@@ -62,4 +72,22 @@ def configure_args(self):
+            ]))
+            args.extend(
+                [
+                    'CFLAGS=' + common_flags,
```
Might have to set `-O2` if it overrides the defaults (haven't verified)

haampie commented on var/spack/repos/builtin/packages/glibc/package.py:
```diff
@@ -192,3 +256,48 @@ def build(self, spec, prefix):
+    @run_after("install")
+    def add_linux_headers(self):
+        cp = which("cp")
+        cp('-r', self.spec['linux-headers'].prefix.include, self.spec.prefix)
```
Spack's `install` can't handle this?

trws replied:
An install function you mean or what?

haampie commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -170,9 +211,9 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
-    depends_on("mpfr@2.4.2:3.1.6", when="@:9.9")
-    depends_on("mpfr@3.1.0:", when="@10:")
-    depends_on("mpc@1.0.1:", when="@4.5:")
+    depends_on("mpfr@2.4.2:3.1.6")
+    depends_on("mpfr@3.1.0:")
+    depends_on("mpc@1.0.1:")
```
Accidental change?

trws replied:
Odd… yes, probably a rebase error or something? Will check, good spotting.

trws replied:
Turns out this was not entirely a mistake.  There's an issue with a newer mpc or mpfr that causes gcc's configure to fail with mpc unable to find libm.so.6 despite having an rpath to it and a -L for the same path.  Using the older verison fixes it, so I'm going to restrict it for just the stages that need it and comment it.

haampie commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -753,30 +837,20 @@ def configure_args(self):
         if spec.satisfies("+binutils"):
             binutils = spec["binutils"].prefix.bin
+            if not os.path.isfile(binutils.ld):
+                binutils = spec["binutils"].prefix.join(self.spec.target_triple).bin
```
Not 100% if this ever worked the right way (as in, absolute paths to the linker). We always have the right ld/as in the path, right?

trws replied:
Yes, but if using an actual cross linker we need to try the cross-name in case it’s an external. We now always build the cross name and link to the unprefixed to help with this so if we require the first stage to always be built with spack this can probably come out.

haampie commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -797,6 +871,97 @@ def configure_args(self):
+                        "--disable-shared",
+                        "--disable-libstdcxx",
+                        "--with-glibc-version=2.38",  # TODO: figure out how to fix this cycle
+                        "--with-native-system-header-dir=/include",
```
Is this basically arbitrary? Normally it points to the dir where glibc is, correct?

trws replied:
It’s actually required, because it’s the path which will be searched *inside the sysroot/-B*. The default is /usr/include I think, which doesn’t work for us.

haampie commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -860,6 +1025,44 @@ def configure_args(self):
+        return (flags, None, None)
+
+    flag_handler = my_flag_handler
+    @run_before("configure", when="+stage1")
```
I still don't fully understand why this is done. Can you leave a comment?

trws replied:
I could swap it to make the method explicit. LSP doesn’t like that we replace a variable with a method, it’s an unsafe override.

haampie commented on var/spack/repos/builtin/packages/gcc/package.py:
```diff
@@ -973,29 +1206,62 @@ def write_rpath_specs(self):
+            name = None
             for line in lines:
+                if add_sysroot:
+                    if loader.search(line):
```
Instead of patching this post-hoc, you can also iirc modify a config file, which is what nix does. That should have the advantage that if you're doing multistage stuff, you can influence gcc in those stages (that may not be relevant here though...)

trws replied:
Oh that would be nice actually, do you have a ref? I couldn’t find where they were doing it for the spec, only saw them wrapping xgcc and using their scripts. We’re already patching configure and a couple others and that might simplify it all.

haampie commented on var/spack/repos/builtin/packages/glibc/package.py:
```diff
@@ -59,6 +63,46 @@ class Glibc(AutotoolsPackage, GNUMirrorPackage):
     version("2.6.1", sha256="6be7639ccad715d25eef560ce9d1637ef206fb9a162714f6ab8167fc0d971cae")
     version("2.5", sha256="16d3ac4e86eed75d85d80f1f214a6bd58d27f13590966b5ad0cc181df85a3493")
 
+    versions = [
```
unused?

trws replied:
Ah, I was going to use that for something and forgot about it. Yeah I think it’s vestigial now.

haampie commented:
I wouldn't mind to force python@3.12 to work around any setup.py related issues.

trws commented:
> Few comments:
> 
> * We could use this approach https://github.com/haampie/spack-intermediate-gcc-example to express the dependencies between the different bootstrap stages, and simplify configuration, so that bootstrapping is a single `make` invocation.
I love this idea! There’s one bug to work around to make it happen I think, stage3 and 4 won’t build with depfile currently, because they get stuck on the Linux-headers compiler being missing despite it already being built.  If we could get that worked around this would be fabulous. 
> * You're using Docker just to get a fixed toolchain to kick off bootstrapping from, right? But in principle it can also be the user's own toolchain? Is the only issue that in Spack you can't differentiate between the just-built gcc and the system toolchain if their versions coincide?

Correct. The first several versions I built with a local toolchain without issue. I added in the docker bit so that the toolchain and os name would be stable, it just simplifies managing the envs. No reason it’s required. Only real requirement is a gcc new and old enough to build the one you’re bootstrapping.

trws commented:
> I wouldn't mind if `python@3.12` was forced to work around any hard-coded system paths `setup.py`.

That's worth a try for sure, though the fix was easy enough I'm at least somewhat in favor of leaving it in there for the moment.

trws committed:
add iconv and remove vestigial versions list from glibc

trws committed:
krb5 needs libedit, only found during a highly parallel build of bootstrap

trws committed:
convert to a depfile/make based bootstrap

bootstraps/Makefile defines the full bootstrap in terms of the
environments in the four subdirectories, after converting to
requirements, this is greatly cleaned up and almost entirely declarative
now.  The remaining issues being having to enforce the glibc hash in
stages 3 and 4, and relying on only the spack built gcc being exactly
version 13.2.0 and the host offering at least one lower-version gcc.

trws committed:
move bootstraps to a more reasonable location

trws committed:
add comment to glibc about linux-headers dep type

trws commented:
@haampie I just realized one challenge we may need to figure out here.  The `gcc-runtime` package in stage 2-4 seems to be re-using ones built with previous compilers.  I'm guessing this is because the paths happen to match or something so the hash works out the same.  It doesn't seem to be causing issues, but might be worth a closer look once we get through the rest.

trws committed:
update stage numbers, fix up makefile

This is getting much closer, full bootstrap works again, stage numbers
line up, depfile all the way through, much cleaner.

trws committed:
perl, allow wrappers to work in perl packages built with spack perl

trws committed:
first step to getting llvm building and working on spackos, builds a working compiler that needs path tweaks to work, fails to build runtimes because of path issues

alalazo commented on share/spack/spackos-bootstrap/stage1/spack.in.yaml:
```diff
@@ -0,0 +1,19 @@
+# configuration settings.
+spack:
+  include:
+    - compilers.yaml
```
Where is this include file supposed to be found? I feel like the spackos-bootstrap directory is missing a `Makefile` or something similar :thinking:

alalazo commented on var/spack/repos/builtin/packages/ncurses/package.py:
```diff
@@ -46,6 +46,11 @@ class Ncurses(AutotoolsPackage, GNUMirrorPackage):
         values=("none", "5", "6"),
         multi=False,
     )
+    variant(
+        "cxx",
+        default=True,
+        description="build c++ bindings",
+    )
```
`ncurses` changes can be extracted in a separate PR

trws replied:
Yup, I still need to extract this.

alalazo commented on var/spack/repos/builtin/packages/sed/package.py:
```diff
@@ -25,6 +25,8 @@ class Sed(AutotoolsPackage, GNUMirrorPackage):
 
     executables = ["^sed$"]
 
+    tags = ["build-tools"]
```
Similarly, we can extract to a separate PR adding the `build-tools tag to existing packages

trws replied:
These are already merged actually, in #42110

alalazo commented on var/spack/repos/builtin/packages/sed/package.py:
```diff
@@ -25,6 +25,8 @@ class Sed(AutotoolsPackage, GNUMirrorPackage):
 
     executables = ["^sed$"]
 
+    tags = ["build-tools"]
```
Similarly, we can extract to a separate PR adding the `build-tools` tag to existing packages

alalazo replied:
Similarly, we can extract to a separate PR adding the `build-tools tag to existing packages

alalazo commented on var/spack/repos/builtin/packages/munge/package.py:
```diff
@@ -43,6 +43,7 @@ def configure_args(self):
         localstatedir = self.spec.variants["localstatedir"].value
         if localstatedir != "PREFIX/var":
             args.append("--localstatedir={0}".format(localstatedir))
+        args.append(f"--with-pkgconfigdir={self.spec.prefix}/lib/pkgconfig")
```
Also this one can possibly be a fix on `munge`

trws replied:
Oh, I forgot about this one, yes this can split out too.  Crazy bug, it only installs a pkgconfig file if you set the target explicitly or there is a /usr/lib/pkgconf directory. 🤷

alalazo commented on var/spack/repos/builtin/packages/mpfr/package.py:
```diff
@@ -46,7 +47,7 @@ class Mpfr(AutotoolsPackage, GNUMirrorPackage):
     )
 
-    force_autoreconf = True
+    # force_autoreconf = True
```
Why is this commented out (alternatively, why was this `True` in the first place)?

trws replied:
I _think_ it was true because of an old config.guess or similar that wasn't playing nice with some architecture.  That doesn't seem to be the case now, and dropping it removes several build dependencies.  I'm not sure it actually matters anymore though, this dates from when I was trying to untangle the C++ bootstrap cycle between libstdc++, gcc and gmp that has since been solved better by building gmp without its c++ bindings in the second stage.  Will try it.

alalazo commented on var/spack/repos/builtin/packages/mpfr/package.py:
```diff
@@ -26,6 +26,7 @@ class Mpfr(AutotoolsPackage, GNUMirrorPackage):
     version("3.1.4", sha256="d3103a80cdad2407ed581f3618c4bed04e0c92d1cf771a65ead662cc397f7775")
     version("3.1.3", sha256="f63bb459157cacd223caac545cb816bcdb5a0de28b809e7748b82e9eb89b0afd")
     version("3.1.2", sha256="79c73f60af010a30a5c27a955a1d2d01ba095b72537dab0ecaad57f5a7bb1b6b")
+    version("2.4.2", sha256="c7e75a08a8d49d2082e4caee1591a05d11b9d5627514e678f02d66a124bcf2ba")
```
This can definitely be added to `mpfr` in `develop as a standalone PR.

alalazo commented on var/spack/repos/builtin/packages/linux-pam/package.py:
```diff
@@ -20,11 +20,11 @@ class LinuxPam(AutotoolsPackage):
-
     def configure_args(self):
-        config_args = ["--includedir=" + self.prefix.include.security]
+        config_args = [
+            "--includedir=" + self.prefix.include.security,
+            # need docbook dep without this
+            "--disable-regenerate-docu",
+            "--disable-doc",
+        ]
```
Another change that can possibly be extracted

alalazo reviewed (commented):
A good chunk of the package changes can be extracted in separate PRs. Same for adding the `build-tools` tag on a few packages.

trws committed:
actually add top-level makefile, thanks @alalazo

haampie commented:
Stage 2 is missing Stage 1's environment view PATH (missing awk, bison etc)

alalazo commented:
It seems to work also on Ubuntu 20.04:

![Screenshot from 2024-01-25 15-36-45](https://github.com/spack/spack/assets/4199709/955ab0a1-3b4b-415b-9cd9-25f809723c18)

haampie commented on share/spack/spackos-bootstrap/stage4/spack.in.yaml:
```diff
@@ -0,0 +1,75 @@
+  # spack bootstrap doesn't work, install manually
+  - gnupg
+  - patchelf
+  - clingo
```
```suggestion
```

haampie commented on share/spack/spackos-bootstrap/stage4/spack.in.yaml:
```diff
@@ -0,0 +1,75 @@
+  - gnupg
+  - patchelf
+  - clingo
+  - clingo-bootstrap
```
```suggestion
  - clingo-bootstrap +optimized
```

trws replied:
This one I have to back out sadly.  For some reason `+optimized` causes an ICE in lto in stage4.  Would be good to see if we can bring it back, but for now I'm backing this out.

trws replied:
Only seemed to happen on aarch64 of the two I'm testing.

haampie commented on share/spack/spackos-bootstrap/stage4/spack.in.yaml:
```diff
@@ -0,0 +1,75 @@
+  # if building outside of spackos, python somehow grabs something out of the
+  # local include paths, build like this:
+  # $ bwrap --dev-bind / / --tmpfs /usr/local --tmpfs /usr/include/x86_64-linux-gnu spack install -v python
+  - python
```
```suggestion
  - python +optimized
```

alalazo commented on share/spack/spackos-bootstrap/stage1/spack.in.yaml:
```diff
@@ -0,0 +1,19 @@
+    - compilers.yaml
+  specs:
+  - binutils +stage1
+  - gcc @13.2.0 +stage1 +binutils ~bootstrap languages=c,c++
```
Not related to this PR, but we have:
```
"LDFLAGS=-static-libstdc++ -static-libgcc "
```
set by `gcc` recipe, which we may want to get rid of?

alalazo replied:
It seems these flags were set back in the days in f8e046bb33b23890e1e3bf92866bd6ea5cefbbec

trws replied:
Wow. That's from a while ago.  It's useful to have those for early stages, though we don't necessarily need them for later ones.

alalazo commented on share/spack/spackos-bootstrap/stage1/spack.in.yaml:
```diff
@@ -0,0 +1,19 @@
+  include:
+    - compilers.yaml
+  specs:
+  - binutils +stage1
```
In my case `binutils` is picking up `sed` and `grep` from the system.

alalazo commented on share/spack/spackos-bootstrap/stage2/spack.in.yaml:
```diff
@@ -0,0 +1,34 @@
+  include:
+    - compilers.yaml
+  specs:
+  - glibc
+  - binutils
+  - gcc@13.2.0 ~bootstrap ~graphite +binutils
+  - gmp
+  # we have no C++ right now, disable
+  - berkeley-db ~cxx ~stl
+  - ncurses ~cxx
```
Generally, most specs are missing build deps on `grep` and `sed`, but we can safely assume those are on the host system I guess. `libxcrypt` and `glibc` rely on `/usr/bin/perl` in my case.

trws replied:
Ah, yes.  The intention is that stage1 should build everything the other stages need, but it's relying on paths getting set, and I just realized some of this was missing.  I'm working on a fix to keep bootstrap working without relying on host stuff as much, but not explicitly adding all of those yet.

alalazo commented on share/spack/spackos-bootstrap/stage4/spack.in.yaml:
```diff
@@ -0,0 +1,75 @@
+  specs:
+  # glibc from stage3, must match the gcc built there or we get two mixed
+  # together, use hash
+  - glibc @GLIBC_HASH@
```
Wondering if we can omit the hash and:
```yaml
packages:
  glibc:
    buildable: false
```

trws replied:
That's a good idea, it seems like it should work, I'll check.  Would be really cool to not need to treat these like templates.

trws replied:
Sadly this does not work.  Setting buildable:false forces spack to treat it as an external, but it isn't.  If I made it one, then it would work but then we have the issue that it's an external.

trws replied:
On an upside, after reworking things to use requirements rather than preferences, it actually doesn't seem necessary to pin the hash anymore.  May end up in the same place anyway. 🤷

trws committed:
Update share/spack/spackos-bootstrap/stage4/spack.in.yaml

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

trws committed:
Update share/spack/spackos-bootstrap/stage4/spack.in.yaml

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

trws commented:
> * Stage 2 is missing Stage 1's environment view PATH (missing awk, bison etc)
I hit this earlier today and was just fixing it, patch up shortly.

>     * `clingo-bootstrap@spack (+optimized)` instead of `clingo` in the final stage
Took this, really good point, thanks!

>     * Further PGO/LTO for last GCC and binutils?

Yes, I really want to do that.  I'm leaving it off for now while I have to iterate over it, but when we get near merging we should absolutely do that.  I'm also tempted to set default lto flags, specifically `-flto=auto -ffat-lto-objects -flto-plugin` IIRC like fedora did so we can get LTO for every library that accepts our injected flags, but the fat objects keep all normal links working as well.  Spend some space, potentially gain a significant amount of LTO benefits.  Still considering how to do it all the way down to libc, might require making stage4 a _second_ glibc build that uses the stage3 LTO capable compiler to build from glibc up.  There are a couple new dependency issues to deal with there but it would be really neat.

trws committed:
fix PATH to point to stage1 view

trws committed:
fix write permission for owner on cpanm

trws committed:
update paths in dockerfile

trws committed:
add add nodep target for the container

trws committed:
clean up stage packages

trws committed:
remove gettext and texinfo from the bottom of the stack by default

trws committed:
fix gcc-runtime behavior by version-naming each compiler

trws commented:
Quick update here:

I've had to focus on other things for the past while, I'm going to try to get this rebased today and at least basically straightened out.  My understanding of the main thing that needs to happen before the overall thing can be merged is that it needs to build, at least the final stage, in CI.  This should require:

* [ ] rebase/cleanup
* [ ] add a stage3 container or build cache built from this somewhere that we can use it as a CI stack base
* [ ] add a stack that builds stage3 again (stage4/stable) with the stage3 container

I'm not entirely sure how to get the image into our container registry, so that's a bit of a blocker, but I'll see what I can do.

Mic92 commented:
Thanks. I was waiting for it: https://github.com/spack/spack/discussions/39712

trws committed:
Merge remote-tracking branch 'upstream/develop' into bootstrap-test

teaguesterling commented:
A few observations after trying this out myself. Frankly, I'm amazed I was able to get this far, and these are just notes to spare others some time. A surprising number of packages worked just fine (noted later)
* GCC doesn't like to build as it gets confused and thinks it's cross-compiling. I haven't had enough time to dig into that yet.
* At least one "base" tool was missing that some packages expected to build: `file`. Was easily solved with a symlink.
* I had to create three empty packages that the solver thought were needed but missing: `nvpl-blas`, `nvpl-lapack`, `opencl-icd-loader`. Not sure what was asking for them but their absence caused issues but it may have been due to a repo I added for testing.
* At least one program is not getting linked properly and needs LD_LIBRARY_PATH set (`exo-open` from Xfce4). 
* Some packages that come with binaries (e.g., cargo in `rust-bootstrap`) need something like `ld_config` to find libraries. You can work around it by cheating with `LD_LIBRARY_PATH=/spack/base/.spack-env/view/lib` for some things.
* Rust still wont work even with the above hack, and frustrating fails around step 4000 of 6000ish.
* LLVM wont compile and fails after an equally long period of time.
* gobject-introspection wont compile due to gir-1.0 "not being a dynamic executable". 
* (as noted below) I had to create a `/lib` and `/lib64` with a few binaries that expected libraries there. 

Other unrelated observations from my testing (all future projects):
 * The introspection (gobject-introspection) variant should probably be sticky due to the number of places it needs to propagate (something I'll probably take on soon).
* Frustratingly: `rust` (because of `librsvg`) AND `llvm` (because of mesa) have become default dependencies for `gtkplus`.
* I was surprised to learn that the `openjdk` package is all pre-compiled binaries and explicitly looks for `/lib64/ld-linux-x86-64.so.2` as well as trying to find `libz.so.1` through ld_config. There is probably some separate work there.
* dbus compiles but wont run with "--system" due to the socket name being too long.

Things that built just fine with either no changes or minor configure/meson flag alterations (not that ANY of this is actually necessary for a spackos):
* libX11 (and related)
* mesa
* glib 
* gtk+
* turbovnc
* Xfce4 (core, terminal, mousepad)

For fun (fonts and icons are still a work-in-progress, obviously)
![image](https://github.com/spack/spack/assets/71129/0d363b41-5454-40ae-909f-f8f6c9503f5f)

Mic92 commented:
If you perform updates, you might need something that can trigger restarts on services, sockets and mounts. Here is an example implementation for systemd: https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/sw/switch-to-configuration-ng/src/main.rs
